go/gas-oracle/oracle/gas_price_oracle.go

```diff
@@ -0,0 +1,216 @@
+package oracle
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/go/gas-oracle/bindings"
+	"github.com/ethereum-optimism/optimism/go/gas-oracle/gasprices"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// errInvalidSigningKey represents the error when the signing key used
+// is not the Owner of the contract and therefore cannot update the gasprice
+var errInvalidSigningKey = errors.New("invalid signing key")
+
+// errNoChainID represents the error when the chain id is not provided
+// and it cannot be remotely fetched
+var errNoChainID = errors.New("no chain id provided")
+
+// errNoPrivateKey represents the error when the private key is not provided to
+// the application
+var errNoPrivateKey = errors.New("no private key provided")
+
+// GasPriceOracle manages a hot key that can update the L2 Gas Price
+type GasPriceOracle struct {
+	chainID   *big.Int
+	ctx       context.Context
+	stop      chan struct{}
+	contract  *bindings.GasPriceOracle
+	backend   DeployContractBackend
+	gasPricer *gasprices.GasPricer
+	config    *Config
+}
+
+// Start runs the GasPriceOracle
+func (g *GasPriceOracle) Start() error {
+	if g.config.chainID == nil {
+		return errNoChainID
+	}
+	if g.config.privateKey == nil {
+		return errNoPrivateKey
+	}
+
+	address := crypto.PubkeyToAddress(g.config.privateKey.PublicKey)
+	log.Info("Starting Gas Price Oracle", "chain-id", g.chainID, "address", address.Hex())
+
+	// Fetch the owner of the contract to check that the local key matches
+	// the owner of the contract. If it doesn't match then nothing can be
+	// accomplished.
+	owner, err := g.contract.Owner(&bind.CallOpts{
+		Context: g.ctx,
+	})
+	if err != nil {
+		return err
+	}
+
+	if address != owner {
+		log.Error("Signing key does not match contract owner", "signer", address.Hex(), "owner", owner.Hex())
+		return errInvalidSigningKey
+	}
+
+	// TODO: Errors in this goroutine should write to an error channel
+	// and be handled externally
+	go g.Loop()
+
+	return nil
+}
+
+func (g *GasPriceOracle) Stop() {
+	close(g.stop)
+}
+
+func (g *GasPriceOracle) Wait() {
+	<-g.stop
+}
+
+func (g *GasPriceOracle) Loop() {
```

Comment:
Nit: I often use the `loop` function in such a way that it has no initialization logic (non-loop) logic. So when I write these functions I make the (mostly stylistic) choice to put things like the `gasPriceUpdater` initialization in the `start(..)` or `new..(..)` function & keep the loop super minimal.

Would love to understand why you used this pattern -- is it more standard? I definitely don't have a lot of context around this so no strong opinions, just trying to sync up our styles for future consistency purposes

Reply:
The alternative would be to have another function that does the initialization and then call loop where loop either accepts a `GasPriceUpdater` or assumes that its set on the `GasOracle`. We just need to be mindful of having functions that set members of the struct outside of the constructor, unless we want to move creation of the `GasPriceUpdater` to the `GasOracle` struct itself

Reply:
Another design patter is putting the goroutine creation inside of the function itself so that its transparent to the caller

Reply:
I've updated this to simplify the loop